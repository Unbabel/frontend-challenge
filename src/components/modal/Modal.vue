<script>
	// Store
	import { mapState, mapActions } from "vuex";

	export default {
		methods: {
			...mapActions({
				toggleModalVisibility: "toggleModalVisibility",
			}),
			closeModal() {
				this.toggleModalVisibility();
			},
		},
		computed: {
			...mapState({
				showModal: "showModal",
			}),
		},
	};
</script>

<template>
	<div id="custom-modal" class="modal" :class="showModal && 'modal-active'">
		<div class="modal-content">
			<div class="flex-row-end">
				<span class="close" @click="closeModal">&times;</span>
			</div>
			<div class="modal-slot-content">
				<slot></slot>
			</div>
		</div>
	</div>
</template>

<style scoped>
	.modal {
		display: none; /* Hidden by default */
		position: fixed;
		z-index: 11;
		left: 0;
		top: 0;
		width: 100%;
		height: 100%;
		overflow: auto;
		background-color: var(--palette-color); /* Fallback color */
		background-color: rgba(0, 0, 0, 0.4); /* Black w/ opacity */
	}

	.modal-active {
		display: block;
	}

	.modal-content {
		background-color: #fefefe;
		margin: 15% auto;
		padding: 1.25rem;
		border: 1px solid #888;
		width: fit-content;
		min-width: 25%;
	}

	.modal-slot-content {
		margin: 0rem 2rem 0.5rem 2rem;
	}

	.flex-row-end {
		display: flex;
		flex-direction: row;
		justify-content: end;
	}

	.close {
		color: #aaa;
		float: right;
		font-size: 1.5rem;
		font-weight: bold;
	}

	.close:hover,
	.close:focus {
		color: black;
		text-decoration: none;
		cursor: pointer;
	}
</style>
