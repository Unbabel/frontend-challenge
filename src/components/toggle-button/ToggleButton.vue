<script setup>
	import { computed } from "vue";

	const { checked, onChangeCallback } = defineProps({
		checked: {
			type: Boolean,
			required: true,
		},
		onChangeCallback: {
			type: Function,
			required: true,
		},
	});

	const isChecked = computed(() => checked);
</script>

<template>
	<div>
		<label class="switch">
			<input type="checkbox" :checked="isChecked" teste="lol" />
			<span
				class="slider"
				@click="onChangeCallback"
				data-test="toggle-button"
			></span>
		</label>
	</div>
</template>

<style scoped>
	.switch {
		position: relative;
		display: inline-block;
		width: 3rem;
		height: 1.5rem;
	}

	.switch input {
		opacity: 0;
		width: 0;
		height: 0;
	}

	.slider {
		position: absolute;
		cursor: pointer;
		top: 0;
		left: 0;
		right: 0;
		bottom: 0;
		background-color: var(--palette-lighter-grey);
		-webkit-transition: 0.4s;
		transition: 0.4s;
		border-radius: 2.125rem;
	}

	.slider:before {
		position: absolute;
		content: "";
		height: 1rem;
		width: 1rem;
		left: 0.25rem;
		bottom: 0.25rem;
		background-color: var(--palette-white);
		-webkit-transition: 0.4s;
		transition: 0.4s;
		border-radius: 50%;
	}

	input:checked + .slider {
		background-color: #2196f3;
	}

	input:checked + .slider:before {
		-webkit-transform: translateX(1.625rem);
		-ms-transform: translateX(1.625rem);
		transform: translateX(1.625rem);
	}
</style>
