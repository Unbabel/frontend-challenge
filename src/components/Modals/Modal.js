import * as React from 'react'
import '../../resources/styles/modal.scss'

const Modal = ({ onClose }) => {
    return (<React.Fragment>
        <div className="modal">
                <div className="modal-content">
                    <span onClick={onClose} className="close">&times;</span>
                    <p>Data Is Uploaded..</p>
                </div>
        </div>
    </React.Fragment>
    )
}
export default Modal;