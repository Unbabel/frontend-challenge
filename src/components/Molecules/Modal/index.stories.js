import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import Modal from '.'

storiesOf('Molecules/Modal', module)
  .add('default', () => (
    <Modal onClose={action('closed')} isOpen>
"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua    </Modal>
  ))
  .add('with title', () => (
    <Modal onClose={action('closed')} title="Hello" isOpen>
"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua    </Modal>
  ))
  .add('closeable', () => (
    <Modal onClose={action('closed')} title="Hello" closeable isOpen>
"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua    </Modal>
  ))
