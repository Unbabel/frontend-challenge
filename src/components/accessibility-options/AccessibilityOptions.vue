<script>
	// Store
	import { mapState, mapActions } from "vuex";

	// Components
	import Modal from "../modal/Modal.vue";
	import ToggleButton from "../toggle-button/ToggleButton.vue";

	// Utils
	import { addOrRemoveClassNameFromElTag } from "../../shared/utils/accessibilityOptions";

	export default {
		components: {
			Modal,
			ToggleButton,
		},
		methods: {
			...mapActions({
				toggleModalVisibility: "toggleModalVisibility",
				toggleAccessibilityOption: "toggleAccessibilityOption",
			}),
			openModal() {
				this.toggleModalVisibility();
			},
			toggleElderMode() {
				addOrRemoveClassNameFromElTag("html", "elder");
				this.toggleAccessibilityOption("elder");
			},
			toggleColorBlindness() {
				this.toggleAccessibilityOption("colorBlindness");
			},
		},
		computed: {
			...mapState({
				showModal: "showModal",
				isElderModeActive: (state) => state.accessibilityOptions.elder,
				isColorBlindnessActive: (state) =>
					state.accessibilityOptions.colorBlindness,
			}),
		},
	};
</script>

<template>
	<div>
		<div class="accessibility-container" aria-label="Accessibility Options">
			<div
				class="accessibility"
				@click="openModal"
				data-test="accessibility-div"
			>
				<img
					src="/src/assets/accessibility.svg"
					alt="accessibility-icon"
					data-test="accessibility-icon"
				/>
			</div>
		</div>
		<Modal>
			<h3>Accessibility Profiles</h3>
			<div class="accessibility-options-grid">
				<p>Elder</p>
				<ToggleButton
					:onChangeCallback="toggleElderMode"
					:checked="isElderModeActive"
				/>
				<p>Color Blindness</p>
				<ToggleButton
					:onChangeCallback="toggleColorBlindness"
					:checked="isColorBlindnessActive"
				/>
			</div>
		</Modal>
	</div>
</template>

<style scoped>
	.accessibility-container {
		position: fixed;
		z-index: 10;
		bottom: 1.25rem;
		left: 0.625rem;
		width: fit-content;
	}
	.accessibility {
		background: var(--palette-purple);
		border-radius: 50%;
		margin: 0.625rem;
		height: fit-content;
		width: fit-content;
		padding: 0.5rem;
		cursor: pointer;

		box-shadow: 0 0 0 0 rgb(129, 94, 194);
		transform: scale(1);
		animation: pulse 2s infinite;
	}

	img {
		height: 2rem;
	}

	@keyframes pulse {
		0% {
			transform: scale(0.95);
			box-shadow: 0 0 0 0 rgba(129, 94, 194, 0.7);
		}

		70% {
			transform: scale(1);
			box-shadow: 0 0 0 10px rgba(142, 68, 173, 0);
		}

		100% {
			transform: scale(0.95);
			box-shadow: 0 0 0 0 rgba(142, 68, 173, 0);
		}
	}

	h3 {
		text-align: center;
	}

	.accessibility-option {
		column-gap: 2rem;
	}

	.accessibility-options-grid {
		display: grid;
		grid-template-columns: 70% 29%;
		align-items: center;
	}
</style>
