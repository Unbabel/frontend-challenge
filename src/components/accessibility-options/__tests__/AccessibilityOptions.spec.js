import Vuex from "vuex";
import { describe, it, expect, beforeEach, vi } from "vitest";
import { mount, createLocalVue, shallowMount } from "@vue/test-utils";

// Components
import AccessibilityOptions from "../AccessibilityOptions.vue";
import Modal from "../../modal/Modal.vue";
import ToggleButton from "../../toggle-button/ToggleButton.vue";

const localVue = createLocalVue();
localVue.use(Vuex);

const store = new Vuex.Store({
	state: {
		showModal: false,
		accessibilityOptions: {
			elder: false,
			colorBlindness: false,
		},
	},
	actions: {
		toggleModalVisibility: () => true,
		toggleAccessibilityOption: () => true,
	},
});

describe("AccessibilityOptions", () => {
	let wrapper;

	beforeEach(() => {
		// Arrange
		vi.restoreAllMocks();
		wrapper = shallowMount(AccessibilityOptions, { localVue, store });
	});

	it("matches snapshot", () => {
		// Assert
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("renders the child components properly", () => {
		// Act
		const modal = wrapper.findComponent(Modal);
		const toggleButtons = wrapper.findAllComponents(ToggleButton);

		// Assert
		expect(modal.exists()).toBe(true);
		expect(toggleButtons.exists()).toBe(true);
		expect(toggleButtons).toHaveLength(2);
	});

	it("renders the accessibility options title", () => {
		// Assert
		expect(wrapper.text()).toContain("Accessibility Profiles");
	});

	it("has correct accessibility options", () => {
		// Assert
		expect(wrapper.text()).toContain("Elder");
		expect(wrapper.text()).toContain("Color Blindness");
		expect(wrapper.text()).not.toContain("Unknown Option");
	});

	it("calls toggleElderMode on toggle button click", () => {
		// Arrange
		const spy = vi.spyOn(AccessibilityOptions.methods, "toggleElderMode");
		wrapper = mount(AccessibilityOptions, { localVue, store });

		// Act
		const toggleButtons = wrapper.findAll('[data-test="toggle-button"]');
		toggleButtons.at(0).trigger("click");

		// Assert
		expect(spy).toBeCalled();
	});

	it("calls toggleColorBlindness on toggle button click", () => {
		// Arrange
		const spy = vi.spyOn(
			AccessibilityOptions.methods,
			"toggleColorBlindness"
		);
		wrapper = mount(AccessibilityOptions, { localVue, store });

		// Act
		const toggleButtons = wrapper.findAll('[data-test="toggle-button"]');
		toggleButtons.at(1).trigger("click");

		// Assert
		expect(spy).toBeCalled();
	});

	it("calls openModal on accessibility button click", () => {
		// Arrange
		const spy = vi.spyOn(AccessibilityOptions.methods, "openModal");
		wrapper = mount(AccessibilityOptions, { localVue, store });

		// Act
		const btn = wrapper.find('[data-test="accessibility-div"]');
		btn.trigger("click");

		// Assert
		expect(spy).toBeCalled();
	});

	[
		{ clx: "accessibility-container" },
		{ clx: "accessibility" },
		{ clx: "accessibility-options-grid" },
	].forEach(({ clx }) =>
		it("has correct css classes", () => {
			// Act
			const divs = wrapper.findAll("div");
			const filteredDiv = divs.filter((div) => div.classes(clx));

			// Assert
			expect(filteredDiv).toHaveLength(1);
		})
	);
});
