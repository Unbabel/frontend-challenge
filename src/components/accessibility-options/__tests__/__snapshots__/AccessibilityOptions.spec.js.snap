// Vitest Snapshot v1

exports[`AccessibilityOptions > matches snapshot 1`] = `
"<div data-v-8789ad59=\\"\\">
  <div data-v-8789ad59=\\"\\" aria-label=\\"Accessibility Options\\" class=\\"accessibility-container\\">
    <div data-v-8789ad59=\\"\\" data-test=\\"accessibility-div\\" class=\\"accessibility\\"><img data-v-8789ad59=\\"\\" src=\\"/src/assets/accessibility.svg\\" alt=\\"accessibility-icon\\" data-test=\\"accessibility-icon\\"></div>
  </div>
  <modal-stub data-v-8789ad59=\\"\\">
    <h3 data-v-8789ad59=\\"\\">Accessibility Profiles</h3>
    <div data-v-8789ad59=\\"\\" class=\\"accessibility-options-grid\\">
      <p data-v-8789ad59=\\"\\">Elder</p>
      <togglebutton-stub data-v-8789ad59=\\"\\" onchangecallback=\\"[Function]\\"></togglebutton-stub>
      <p data-v-8789ad59=\\"\\">Color Blindness</p>
      <togglebutton-stub data-v-8789ad59=\\"\\" onchangecallback=\\"[Function]\\"></togglebutton-stub>
    </div>
  </modal-stub>
</div>"
`;
